Add capabilities to disable podman socket forwarding and podman-ready notifications from the guest

cfergeau opened the pull request:
These podman-machine feature are not needed by macadam, and are actually detrimental to our goal of running unmodified cloud images. This work will allow to disable this.

cfergeau committed:
machine: Make podman socket forwarding optional

podman-machine is forwarding some podman-specific sockets from the VM to
the host, which requires ssh access to the VM and the use of gvproxy.
This is not needed by macadam, and is currently detrimental as we want
to allow to run unmodified VM images.
This commit adds a new `Capabilities` struct to the vmconfigs, which
will be used for such optional features. podman-machine will enable
socket forwarding, but macadam will be able to opt out of it, and the
podman-machine code is modified to only try to forward sockets when the
capability is enabled.

cfergeau committed:
applehv: Fix comment location

Move comment next to the code it applies to instead of having an extra
blank line between comment and code.

cfergeau committed:
machine: Make "ready" signal optional

podman-machine currently adds a `podman-ready` systemd unit to its
virtual machine. This unit sends a signal to the host to indicate the
guest OS has finished booting.
This is unwanted for macadam which aims at running unmodified VM images.
This commit uses the Capabilities mechanism to make this podman-machine
feature optional.

cfergeau committed:
sockets: Remove unused functions

lstocchi commented on pkg/machine/vmconfigs/machine.go:
```diff
@@ -181,11 +181,14 @@ func (mc *MachineConfig) Remove(machines map[string]bool, saveIgnition, saveImag
 		apiSocket.GetPath(),
 		logPath.GetPath(),
 	}
+	if mc.Capabilities.HasReadyUnit {
```
This caused a failure to me bc I already had a machine on my laptop. When trying to delete it to init another one, it failed bc mc.Capabilities is nil

cfergeau replied:
I added some `GetHasReadyUnit()` methods to deal with this, however this early in macadam development, it is in my opinion acceptable if the upgrade path is not smooth (ie if the VM needs to be removed with the old binary before upgrading)

lstocchi replied:
My concern is if one day podman will use our stuff, everything will break up for them bc there is no Capabilities data, and the user is forced to recreate the podman machines. Atleast if we put a guard we are safe in any case.
I'm ok with the `GetHasReadyUnit()`... maybe the `if` part should return true? If there is no capabilities (`if caps == nil {`) specified i assume it is an old (podman??) machine and should use the old code. WDYT?

cfergeau replied:
> My concern is if one day podman will use our stuff, everything will break up for them bc there is no Capabilities data, and the user is forced to recreate the podman machines. Atleast if we put a guard we are safe in any case.

Ah good point!

> I'm ok with the GetHasReadyUnit()... maybe the if part should return true? If there is no capabilities (if caps == nil {) specified i assume it is an old (podman??) machine and should use the old code. WDYT?

Yes, this was my intent but I messed that up :) I’ll fix it.

lstocchi commented on pkg/machine/vmconfigs/machine.go:
```diff
@@ -195,8 +198,10 @@ func (mc *MachineConfig) Remove(machines map[string]bool, saveIgnition, saveImag
 		var errs []error
-		if err := connection.RemoveConnections(machines, mc.Name, mc.Name+"-root"); err != nil {
-			errs = append(errs, err)
+		if mc.Capabilities.ForwardSockets {
```
same as above

lstocchi commented on pkg/machine/vmconfigs/machine.go:
```diff
@@ -209,8 +214,10 @@ func (mc *MachineConfig) Remove(machines map[string]bool, saveIgnition, saveImag
 		}
-		if err := readySocket.Delete(); err != nil {
-			errs = append(errs, err)
+		if mc.Capabilities.HasReadyUnit {
```
same as above

lstocchi commented on pkg/machine/shim/host.go:
```diff
@@ -112,6 +118,8 @@ func Init(opts machineDefine.InitOptions, mp vmconfigs.VMProvider) error {
 
 	mc.Version = vmconfigs.MachineConfigVersion
+	mc.Capabilities = opts.Capabilities
+	mc.Capabilities.ForwardSockets = false
```
Should this be removed?

lstocchi reviewed (commented):
Overall LGTM, just the comments above which are mandatory bc you cannot clean old VMs otherwise. The same remarks can be done to other parts as you could start an old VM which would have the Capabilities nil

cfergeau committed:
wsl: Disable socket forwarding

This does more than just disabling forwarding of the podman socket, this
also disables the "linger" configuration, and podman-specific
configuration such as registries.
`ForwardSockets` could be renamed to something more generic such as
`SetupPodman`

cfergeau commented:
I’ve kept the `fixup` commits separate for now to make it easier to see what changed, I’ll squash them.

lstocchi reviewed (approved):
Tested on Mac and Win. Works perfectly. Great job!! LGTM 🚀
